### ISSUE TITLE
Custom axes labeling with empty chart are drawn on top of each other

### ISSUE THREAD

**Issue by mattijn** [vega/vega#812]
Once a chart with custom axes labeling is defined without data, the custom axes labeling are drawn on top of each other at the origin of the chart bottom left. Like this:

![customlabeling_nodata](https://cloud.githubusercontent.com/assets/5186265/26111787/5d85fcd8-3a56-11e7-9fd3-ef08bcf3074a.PNG)

Once the data is requested the chart is displayed properly. Eg. like this:
![customlabeling_data](https://cloud.githubusercontent.com/assets/5186265/26111908/b78c9994-3a56-11e7-8164-d7a1076fa65a.PNG)

Is there a suggested route how to avoid this behavior?

The empty chart is drawn with the following code

```
{
  "$schema": "http://vega.github.io/schema/vega/v3.0.json",  
  "height": 307,
  "width":600,
  "padding": { "left": 35, "top": 40, "bottom": 30, "right": 30 },
  "autosize": "pad",
  "data": [
    {
      "name": "source",
      "values": [{}],
      "format": { 
        "type": "json",
        "parse": {
          "frequentie": "number",
          "waarde": "number"
        }
      },
      "transform": [
        {
          "type": "filter",
          "expr": "datum[\"frequentie\"] !== 0.0 && datum[\"frequentie\"] !== null && !isNaN(datum[\"frequentie\"]) && datum[\"waarde\"] !== null && !isNaN(datum[\"waarde\"])"
        },
        {
          "type": "collect",
          "sort": {
            "field": "frequentie",
            "order": "descending"
          }
        }
      ]
    },
    {
      "name": "layout",
      "values": [
        {
          "_id": 50945,
          "width": 200,
          "height": 200
        }
      ],
      "transform": [
        {
          "type": "formula",
          "as": "width",
          "expr": "width"
        },
        {
          "type": "formula",
          "as": "height",
          "expr": "height"
        }
      ]
    }
  ],
  "marks": [
   {
      "name": "main",
      "type": "group",
      "description": "Frequentielijnen",
      "from": { "data": "layout" },
      "encode": {
        "update": {
          "width": { "field": "width" },
          "height": { "field": "height" }
        }
      },
      "marks": [
        {
          "name": "pathgroup",
          "type": "group",
          "from": {
            "facet": {
              "name": "faceted-path-source",
              "data": "source",
              "groupby": ["berekeningid"]
            }
          },
          "encode": {
            "update": {
              "width": { "field": { "group": "width" } },
              "height": { "field": { "group": "height" } }
            }
          },
          "marks": [
            {
              "name": "marks",
              "type": "line",
              "interactive": false,
              "from": { "data": "faceted-path-source" },
              "encode": {
                "update": {
                  "x": { "scale": "x", "field": "frequentie" },
                  "y": { "scale": "y", "field": "waarde" },
                  "strokeWidth": { "value": 2 },
                  "opacity": [
                    
                    { "value": 1 }
                  ],
                  "stroke": [
                    { "scale": "color", "field": "berekeningid" }
                  ],
                  "fill": { "value": "transparent" }
                }
              }
            },
            {
              "type": "symbol",
              "from": { "data": "faceted-path-source" },
              "encode": {
                "update": {
                  "x": { "scale": "x", "field": "frequentie" },
                  "y": { "scale": "y", "field": "waarde" },
                  "align": { "value": "center" },
                  "baseline": { "value": "middle" },
                  "size": { "value": 20 },
                  "strokeWidth": { "value": 3 },
                  "strokeOpacity": { "value": 1.0 },
                  "fillOpacity": { "value": 1.0 },
                  "opacity": [                    
                    { "value": 1.0 }
                  ],
                  "stroke": [                    
                    { "value": "green" }
                  ],
                  "fill": { "value": "transparent" }
                },
                "hover": {
                  "strokeWidth": { "value": 4 },
                  "strokeOpacity": { "value": 1.0 },
                  "fillOpacity": { "value": 1.0 },
                  "opacity": [                    
                    { "value": 1.0 }
                  ],
                  "stroke": [                     
                     { "value": "red" }
                  ],
                  "fill": { "value": "transparent" }
                }
              }
            }
          ]
        }
      ],
      "scales": [
        {
          "name": "x",
          "type": "log",
          "domain": { "data": "source", "field": "frequentie" },
          "range": "width",
          "round": true,
          "reverse": "true",
          "nice": true,
          "zero": false
        },
        {
          "name": "y",
          "type": "linear",
          "domain": { "data": "source", "field": "waarde" },
          "range": "height",
          "round": true,
          "nice": true,
          "zero": false
        },
        {
          "name": "color",
          "type": "ordinal",
          "domain": { "data": "source", "field": "berekeningid", "sort": false },
          "range": ["#e51c23", "#e91e63", "#9c27b0", "#673ab7", "#3f51b5", "#5677fc", "#03a9f4", "#00bcd4", "#009688", "#259b24", "#8bc34a", "#cddc39", "#ffeb3b", "#ffc107", "#ff9800", "#ff5722"]
        },
        {
          "name": "scale_convert",
          "type": "threshold",
          "domain": ["1e-10", "1e-9", "1e-8", "1e-7", "1e-6", "1e-5", "1e-4", "1e-3", "1e-2", "1e-1", "1e+0", "1e1", "1e2", "1e3", "1e4", "1e5", "1e6", "1e7", "1e8", "1e9", "1e10"],
          "range": ["10\u207b\u00b9\u2071", "10\u207b\u00b9\u2070", "10\u207b\u2079", "10\u207b\u2078", "10\u207b\u2077", "10\u207b\u2076", "10\u207b\u2075", "10\u207b\u2074", "10\u207b\u00b3", "10\u207b\u00b2", "10\u207b\u00b9", "10\u2070", "10\u00b9", "10\u00b2", "10\u00b3", "10\u2074", "10\u2075", "10\u2076", "10\u2077", "10\u2078", "10\u2079"]
        }
      ],
      "axes": [
        {
          "scale": "x",
          "format": "e",
          "orient": "bottom",
          "tickCount": 5,
          "tickSize": 6,
          "title": "xlabel (unit)",
          "titlePadding": -4,
          "offset": 10,
          "zindex": 0,
          "values": [1e+0, 1e-1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6],
          "encode": {
            "ticks": {
              "update": {
                "stroke": { "value": "#000" },
                "strokeWidth": { "value": 0.5 }
              }
            },
            "labels": {
              "interactive": true,
              "update": {
                "fill": { "value": "#666666" },
                "angle": { "value": 0 },
                "align": { "value": "right" },
                "font": { "value": "Arial" },
                "dx": { "value": 15 },
                "fontSize": { "value": 12 },
                "text": { "scale": "scale_convert", "field": "value" }
              }
            },
            "title": {
              "update": {
                "font": { "value": "Arial" },
                "fontSize": { "value": 15 },
                "fontWeight": { "value": "normal" },
                "fill": { "value": "#666666" }
              }
            },
            "domain": {
              "update": {
                "stroke": { "value": "#000" },
                "strokeWidth": { "value": 0.5 }
              }
            }
          }
        },
          {
            "scale": "x",
            "domain": false,
            "format": "s",
            "labels": false,
            "grid": true,
            "orient": "bottom",
            "ticks": true,
            "tickSize": 3,
            "tickCount": 6,
            "offset": 10,
            "zindex": 0,
            "encode": {
              "ticks": {
                "update": {
                  "stroke": { "value": "#000" },
                  "strokeWidth": { "value": 0.5 }
                }
              },
              "grid": {
                "update": {
                  "stroke": { "value": "#e6e6e6" },
                  "strokeWidth": { "value": 0.5 }
                }
              }
            }
          },
          {
            "scale": "y",
            "format": "s",
            "orient": "left",
            "tickCount": 8,
            "tickSize": 5,
            "title": "ylabel (unit)",
            "titlePadding": 10,
            "offset": 10,
            "zindex": 1,
            "encode": {
              "ticks": {
                "update": {
                 "stroke": { "value": "#000" },
                  "strokeWidth": { "value": 0.5 }
                }
              },
              "labels": {
                "interactive": true,
                "update": {
                  "fill": { "value": "#666666" },
                  "angle": { "value": 0 },
                  "font": { "value": "Arial" },
                  "fontSize": { "value": 12 },
                  "dx": { "value": -3 }
                }
              },
              "title": {
                "update": {
                  "font": { "value": "Arial" },
                  "fontSize": { "value": 15 },
                  "fontWeight": { "value": "normal" },
                  "fill": { "value": "#666666" }
                }
              },
              "domain": {
                "update": {
                  "stroke": { "value": "#000" },
                  "strokeWidth": { "value": 0.5 }
                }
              }
            }
          },
          {
            "scale": "y",
            "domain": false,
            "format": "s",
            "labels": false,
            "grid": true,
            "orient": "left",
            "ticks": true,
            "tickCount": 8,
            "offset": 10,
            "zindex": 0,
            "gridScale": "x",
            "encode": {
              "ticks": {
                "update": {
                  "stroke": { "value": "#cccccc" },
                  "strokeWidth": { "value": 0.8 }
                }
              },
              "grid": {
                "update": {
                  "stroke": { "value": "#e6e6e6" },
                  "strokeWidth": { "value": 0.5 }
                }
              }
            }
          }
      ]
    }
  ]
}
```

and the `source` data object can be filled with the following data values to get the above displayed chart:

```
{ "waarde": 15.2, "frequentie": 0.1131702, "code": "Waterstand", "berekeningid": 0 },
{ "waarde": 16.8, "frequentie": 0.0069746193, "code": "Waterstand", "berekeningid": 0 },
{ "waarde": 18.3, "frequentie": 0.0000068735044, "code": "Waterstand", "berekeningid": 0 },
{ "waarde": 16.2, "frequentie": 0.1131702, "code": "HBN", "berekeningid": -1 },
{ "waarde": 17.6, "frequentie": 0.011218123, "code": "HBN", "berekeningid": -1 },
{ "waarde": 19.3, "frequentie": 0.0000068735044, "code": "HBN", "berekeningid": -1 },
{ "waarde": 11.9, "frequentie": 0.024304695, "code": "Significante golfhoogte", "berekeningid": -2 },
{ "waarde": 12.1, "frequentie": 0.0026245182, "code": "Significante golfhoogte", "berekeningid": -2 },
{ "waarde": 13.4, "frequentie": 0.000046522305, "code": "Significante golfhoogte", "berekeningid": -2 }
```

**Comment by jheer** [comment]
The axis determines the actual range from the scale, the label values are then simply placed relative to that scale. So, if you have a desired "hardwired" domain for your data you can simply update your scale definition to use a literal domain array rather than have the domain computed from input data.

That is, replace `"domain": {"data": "source", "field": "frequentie"}` with `"domain": [1e+0, 1e-1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6]`.

**Comment by mattijn** [comment]
It will work in this case (and I checked and it probably will as well in my application), but I see it as a workaround. 
Once I delete the following point locations of my dummy values
`{ "waarde": 15.2, "frequentie": 0.1131702, "code": "Waterstand", "berekeningid": 0 },` and `{ "waarde": 16.2, "frequentie": 0.1131702, "code": "HBN", "berekeningid": -1 },` a new [smaller] domain will be computed from my input data, with correct* custom label values. But if I hardwire my domain it will stay the same.

In other words, with my code I had a maximum domain range, _but_ the domain could become smaller depending on the input data.
In the proposed solution I will have a fixed domain range no matter the input data. 

*. I just notice that the label value is still drawn, but now outside the plot
![customlabeling_smallerdomain](https://cloud.githubusercontent.com/assets/5186265/26141987/d76fcf0a-3ade-11e7-88a1-2388c80aa40b.PNG)

**Comment by mattijn** [comment]
In fact, quite happy with the solution offered. Will close issue
